src/idl/src/tree.c

```diff
@@ -1037,14 +1071,78 @@ idl_create_struct(
   return ret;
 }
 
+static void delete_struct_forward_decl(void *ptr)
+{
+  idl_forward_decl_t *node = ptr;
+  idl_delete_name(node->name);
+  free(node);
+}
+
+static const char *describe_struct_forward_decl(const void *ptr)
+{
+  (void)ptr;
+  assert(idl_mask(ptr) & (IDL_STRUCT | IDL_FORWARD));
+  return "struct forward declaration";
+}
+
+idl_retcode_t
+idl_forward_decl_struct(
+  idl_pstate_t *pstate,
+  const idl_location_t *location,
+  idl_name_t *name,
+  void *nodep)
+{
+  idl_retcode_t ret;
+  idl_forward_decl_t *node;
+  idl_declaration_t *declaration;
+  static const size_t size = sizeof(*node);
+  static const idl_mask_t mask = IDL_STRUCT | IDL_FORWARD;
+  static const struct methods methods = {
+    delete_struct_forward_decl, 0, describe_struct_forward_decl };
+  static const enum idl_declaration_kind kind = IDL_SPECIFIER_FORWARD_DECLARATION;
+
+  if ((ret = create_node(pstate, size, mask, location, &methods, &node)))
+    goto err_node;
+  node->name = name;
+  node->type_mask = IDL_STRUCT;
+  if ((ret = idl_declare(pstate, kind, name, node, NULL, &declaration)))
+    goto err_declare;
+  node->scoped_name = declaration->scoped_name;
+  *((idl_forward_decl_t **)nodep) = node;
+  return ret;
+err_declare:
+err_node:
+  return ret;
+}
+
+bool idl_is_forward(const void *ptr)
+{
+  const idl_forward_decl_t *node = ptr;
+
+  if (!(idl_mask(node) & IDL_FORWARD))
+    return false;
+  /* a forward declaration must have a name */
+  assert(node->name && node->name->identifier);
+  return true;
+}
+
+idl_mask_t idl_forward_mask(const void *ptr)
```

Comment:
This doesn't return an `idl_mask_t`, but an `idl_type_t`. But, with the `type_mask` member possibly being redundant, I bet you can simply replace `idl_forward_mask` calls with `idl_type` ones(?)

Reply:
Good point, this is fixed in c80f50d21d8712247a7a6c7c48aae7823ce6dc0a